### PR TITLE
[Associated type inference] De-duplicate solutions as early as possible.

### PR DESCRIPTION
**Pull request by DougGregor**
Associated type inference can end up visiting a large number of potential
solution states, which it ends up having to deduplicate. Perform the
de-duplication step earlier, before we do extensive checking of the
solution itself, to eliminate a bunch of extra work.

Addresses the compile-time regression in rdar://problem/37165620.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test and merge

**Comment by DougGregor** [comment]
@swift-ci please smoke test compiler performance

**Review by graydon — APPROVED** [review]
Thanks for getting on this so quickly!

Given my rough understanding of both the performance bottleneck you've described here (calling checkCurrentTypeWitness before we've bailed out on duplicates) and the structure of the loop, it looks sensible enough to me to move the requisite code up in the loop and move the call down to after the bailout.

I'll queue up a task in my todo list to change these statistics to always-available unless you think they're too much of a corner case to bother tracking.

File: lib/Sema/TypeCheckProtocolInference.cpp [hunk 1]
```diff
@@ -1204,27 +1200,54 @@ void AssociatedTypeInference::findSolutionsRec(
...
-      // We found the same solution. There is no point in recording
-      // a new one.
-      if (sameSolution)
-        return;
+      return true;
+    };
+
+    // If we've seen this solution already, bail out; there's no point in
+    // checking further.
+    if (std::find_if(solutions.begin(), solutions.end(), matchesSolution)
```
Comment (huonw) [inline lib/Sema/TypeCheckProtocolInference.cpp]: FWIW, I think these could probably be `llvm::all_of(solutions, matchesSolution)`.

Comment (DougGregor) [reply lib/Sema/TypeCheckProtocolInference.cpp]: You mean `llvm::any_of`, I presume. I didn't know about that one, thanks!

**Pull request merged**